### PR TITLE
Add support for RISCV hardware performance counter.

### PR DESCRIPTION
**Pull request by NandkumarJoshi**
Note: mcountinhibit csr is not yet accessible. As per latest RiscV privileged spec v1.11, mcountinhibit controls which of the counters increment. Unused counters can be disabled to reduce power consumption. APIs to start / stop counters can be added, once this csr is supported.

### CODE REVIEW

File: src/hwperfmon.c [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+
+/* Return codes */
+#define METAL_HWPERFMON_RET_OK 0
+#define METAL_HWPERFMON_RET_NOK 1
+
+/* Module enable flag */
+static int hwperfmon_enable = 0;
+
+/* Available hardware counters */
+static metal_hwperfmon_counter hwperfmon_count = 0;
```
Comment (bsousi5) [inline src/hwperfmon.c:75]: These global variable won't work for MC designs.

File: src/hwperfmon.c [hunk 2]
```diff
@@ -0,0 +1,325 @@
+/* Copyright 2020 SiFive, Inc */
+/* SPDX-License-Identifier: Apache-2.0 */
+
+#include <metal/hwperfmon.h>
+#include <stdint.h>
+
+/* Macro to generate code within a switch case */
+#define METAL_HWPERFMON_HANDLE_SWITCH(m)                                       \
```
Comment (bsousi5) [inline src/hwperfmon.c:8]: Some funky macro.  Not such a big fan of them, but don't have an alternative at the moment.

**Comment by bsousi5** [comment]
Adding @keith-packard to review for his opinion and view

File: src/hwperfmon.c (hunk 1 above)
Comment (nategraff-sifive) [reply src/hwperfmon.c:75]: Any per-hart storage should be generated by `freedom-metal_header-generator`, in this case probably as members of the RISC-V cpu struct.

Comment (NandkumarJoshi) [reply src/hwperfmon.c:75]: Added checks to make sure only single hart now writes to global variables.

File: src/hwperfmon.c [hunk 3]
```diff
@@ -0,0 +1,356 @@
...
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmevent" #x : "=r"(val));             \
+        break;
+
+/* Macro to read HW performance monitor counter values */
+#if __riscv_xlen == 32
+#define METAL_HWPERFMON_GET_COUNT_REG(x)                                       \
+    case METAL_HWPERFMON_COUNTER_##x:                                          \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x "h" : "=r"(vh));        \
+        __asm__ __volatile__("csrr %0, mhpmcounter" #x : "=r"(vl));            \
```
Comment (keith-packard) [inline src/hwperfmon.c:41]: Is this magically atomic somehow? If not, then an overflow from the low 32 bits into the high 32 bits can cause
the wrong value to be returned. The usual fix for this is to use a loop which reads high, reads low, reads high until the two high reads return the same value:
```
uint32_t h1, l, h2;
do {
   h1 = read_hi();
   l = read_lo();
   h2 = read_hi();
} while (h1 != h2);
```

File: src/hwperfmon.c (hunk 1 above)
Comment (keith-packard) [reply src/hwperfmon.c:75]: How will that help? Are the hwperf enable bits shared by all harts?

File: src/hwperfmon.c [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+/* Available hardware counters */
+static metal_hwperfmon_counter hwperfmon_count = 0;
+
+/* Atomic access flag, place it into ITIM. */
+static int flag __attribute__((aligned(4))) __attribute__((section(".itim"))) =
+    0;
+
+__attribute__((naked)) static int metal_hwperfmon_cas(int *ptr, int old,
+                                                      int new) {
+    /* Perform atomic compare and swap */
```
Comment (keith-packard) [inline src/hwperfmon.c:83]: Can you use the GCC atomic intrinsics instead of open-coding in assembly?

File: src/hwperfmon.c [hunk 5]
```diff
@@ -0,0 +1,356 @@
...
+                         "li a0, 1 # Set return to failure. \n\t"
+                         "jr ra # Return. \n\t");
+#endif
+    /* Return error, if no support */
+    __asm__ __volatile__("li a0, 1 \n\t jr ra \n\t");
+    return 1;
+}
+
+int metal_hwperfmon_enable(void) {
+    /* Make sure only a single hart executes this code */
```
Comment (keith-packard) [inline src/hwperfmon.c:101]: Do we need to enable atomics on all harts, or is it sufficient to enable them on a single hart?

File: src/hwperfmon.c (hunk 1 above)
Comment (NandkumarJoshi) [reply src/hwperfmon.c]: My bad, i mistakenly assumed that these counters are shared between harts. hwperfmon_count is now moved into metal cpu device instance, as suggested by Nate.

File: src/hwperfmon.c (hunk 3 above)
Comment (NandkumarJoshi) [reply src/hwperfmon.c]: Corrected.

File: src/hwperfmon.c (hunk 4 above)
Comment (NandkumarJoshi) [reply src/hwperfmon.c]: Removed related code.

File: src/hwperfmon.c (hunk 5 above)
Comment (NandkumarJoshi) [reply src/hwperfmon.c]: Fixed. All harts can enable their counters.

**Comment by NandkumarJoshi** [comment]
> I have questions about how to reliably read counters on 32-bit systems, and whether we need to enable them on all HARTs in the system.
Reading 'high' twice seems to be reliable. Yes, we need to enable them on all harts.

**Comment by NandkumarJoshi** [comment]
> I would preferred HWPM or HPM or PM, instead of HWPERFMON for all function-names, file-names, and enum/macros.
> Also metal_hwperfmon_enable() should be metal_hwperfmon_init().
> Otherwise LGTM

Fixed.

**Pull request merged**